controllers/openstackdataplanenode_controller.go

```diff
@@ -107,37 +106,17 @@ func (r *OpenStackDataPlaneNodeReconciler) Provision(ctx context.Context, instan
 	return nil
 }
 
-// Inventory struct
-// TODO: make use of the struct below
-//
-//nolint:unused
-type Inventory struct {
-	all struct {
-		hosts struct {
-			host struct {
-				hostVar struct {
-					hostVarValue string
-				}
-			}
-		}
-	}
-}
-
 // GenerateInventory yields a parsed Inventory
 func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode) error {
 	var err error
 
-	inventory := make(map[string]map[string]map[string]map[string]string)
-	all := make(map[string]map[string]map[string]string)
-	host := make(map[string]map[string]string)
-	hostVars := make(map[string]string)
-	hostVars["ansible_host"] = instance.Spec.Node.HostName
-	hostVars["ansible_user"] = instance.Spec.Node.AnsibleUser
-	hostVars["ansible_port"] = strconv.Itoa(instance.Spec.Node.AnsiblePort)
-	host[instance.Name] = hostVars
-	all["hosts"] = host
-	inventory["all"] = all
-
+	var inventory = `all:
```

Comment:
In general I think it's better to build a data structure and then have a function render it into proper yaml. I don't think formatting a string by hand into proper yaml is a pattern we should promote.

Reply:
If I make this change:
```
[fultonj@osp-storage-01 dataplane-operator]$ git diff controllers/openstackdataplanenode_controller.go
diff --git a/controllers/openstackdataplanenode_controller.go b/controllers/openstackdataplanenode_controller.go
index cf04365..00ed9d3 100644
--- a/controllers/openstackdataplanenode_controller.go
+++ b/controllers/openstackdataplanenode_controller.go
@@ -162,6 +162,7 @@ func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context
                cm.Data = map[string]string{
                        "inventory": string(invData),
                }
+               fmt.Printf("inv: %s", string(invData))
                return nil
        })
        if err != nil {
[fultonj@osp-storage-01 dataplane-operator]$ 

```
Then I see the inventory:
```
inv: all:
  hosts:
    openstackdataplanenode-sample:
      ansible_host: openstackdataplanenode-sample.localdomain
      ansible_port: "0"
      ansible_user: ""
```
when I run the controller directly and run:
```
oc kustomize config/samples/ | oc apply -f -
```
So I see my port isn't getting set to 22 as it might be for the role but I assume that will be addressed when we have inheritance working. 

I'm not sure (short commit message) what this patch is trying to accomplish.

Reply:
ya so i would either perefer to build the data structure and then render it too or use go templates to build the inventory file.

Reply:
sorry about the commit message, my idea was just replacing the maps

Reply:
I'm googling around, marshalling & unmarshalling structs with dynamic keys is painful. Will we use the inventory struct elsewhere? If not, I believe go templates will probably be the best option then

Reply:
> "I'm googling around, marshalling & unmarshalling structs with dynamic keys is painful."

Interesting. Would you mind sharing what you found? Is it just the amount of code required or something else?

Reply:
the dynamic key would be the hostname
```yaml
all:
  hosts:
    << dynamic host name  >>:
        hostvars
```        
this is what I found:
https://stackoverflow.com/questions/32147325/how-to-parse-yaml-with-dynamic-key-in-golang


the https://github.com/relex/aini looks interesting, I'll do some experiments

Reply:
but the hostname is just the key in a map so 
my type 

type Group struct {
GroupVars map[string]interface{}
Hosts list[Host]
}

should actually be

type Group struct {
GroupVars map[string]interface{}
Hosts map[string]Host
}

where the key is just the hostname

Reply:
by the way we never have to parse this in go.

we will only ever need to write this so it make it a little simpler but modelling the dictionary filed as maps of sting to structs should make the hostnames simple to handle
if we did need to parse this.

Reply:
thank you @SeanMooney 
I've followed your tips, and tested with:
```
apiVersion: dataplane.openstack.org/v1beta1
kind: OpenStackDataPlaneNode
metadata:
  name: openstackdataplanenode-sample
spec:
  node:
    hostName: openstackdataplanenode-sample.localdomain
    ansibleHost: 192.168.122.18
    ansibleUser: rhosp
    ansiblePort: 80
```

and got:
```
kubectl get configmap dataplanenode-openstackdataplanenode-sample-inventory -o yaml
apiVersion: v1
data:
  inventory: |
    groups:
      all:
        hosts:
          openstackdataplanenode-sample:
            vars:
              ansible_host: openstackdataplanenode-sample.localdomain
              ansible_port: "80"
              ansible_user: rhosp
kind: ConfigMap
metadata:
  creationTimestamp: "2023-02-02T15:42:35Z"
  name: dataplanenode-openstackdataplanenode-sample-inventory
  namespace: default
  resourceVersion: "1307"
  uid: d9377b17-bb57-47f8-ad53-6751addc3edf

```

Reply:
The ansible inventory structure has some quirks so probably need custom (un)marshalling logic to get this perfect, but for now something like this might work https://go.dev/play/p/92S1JDkHGDW

Reply:
ya that looks pretty good